File: examples/java/cases/tasks/task_list/TaskListUiCase.java
package cases.tasks.task_list;

import core.BaseUiCase;
import core.shared.AppBaseContext.AppLogger;
import core.shared.AppInfraContracts.AppHttpClient;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class TaskListUiCase extends BaseUiCase<TaskListUiCase.ViewState> {
  public interface DesignSystemContract {
    String feedback(String type, String message);
    String taskBoard(String columnsHtml);
    String taskColumn(String title, int count, String itemsHtml);
    String taskCard(String title, String description, String status, String actionsHtml);
    String emptyColumnState(String message);
  }

  @FunctionalInterface
  public interface RenderCardActions {
    String render(TaskListDomainCase.Task task) throws Exception;
  }

  public static class ViewState {
    public List<TaskListDomainCase.Task> tasks = List.of();
    public boolean loading;
    public String error;
  }

  public static class ColumnViewModel {
    public String status;
    public String title;
    public List<TaskListDomainCase.Task> tasks = List.of();
    public String emptyMessage;
  }

  public static class ViewModel {
    public List<ColumnViewModel> columns = List.of();
    public String feedback;
  }

  public TaskListUiCase(UiContext ctx) {
    super(ctx, new ViewState());
  }

  @Override
  public String view() throws Exception {
    DesignSystemContract designSystem = resolveDesignSystem();
    ViewState currentState = new ViewState();
    currentState.loading = true;

    try {
      TaskListDomainCase.TaskListOutput result = load(new TaskListDomainCase.TaskListInput());
      currentState.tasks = result.tasks();
      currentState.loading = false;
    } catch (Exception error) {
      currentState.tasks = List.of();
      currentState.loading = false;
      currentState.error = error.getMessage();
    }

    ViewModel model = _viewmodel(currentState);
    RenderCardActions renderCardActions = resolveRenderCardActions();

    StringBuilder columnsHtml = new StringBuilder();
    for (ColumnViewModel column : model.columns) {
      StringBuilder itemsHtml = new StringBuilder();
      if (!column.tasks.isEmpty()) {
        for (TaskListDomainCase.Task task : column.tasks) {
          String actionsHtml = renderCardActions == null ? "" : renderCardActions.render(task);
          itemsHtml.append(designSystem.taskCard(
              task.title(),
              task.description(),
              task.status(),
              actionsHtml
          ));
        }
      } else {
        itemsHtml.append(designSystem.emptyColumnState(column.emptyMessage));
      }

      columnsHtml.append(
          designSystem.taskColumn(column.title, column.tasks.size(), itemsHtml.toString())
      );
    }

    return designSystem.feedback("error", model.feedback)
        + designSystem.taskBoard(columnsHtml.toString());
  }

  @Override
  public void test() throws Exception {
    UiContext testContext = new UiContext();
    testContext.correlationId = "task-list-ui-test";
    testContext.logger = new NoopLogger();
    testContext.api = config -> Map.of(
        "tasks", List.of(
            Map.of(
                "id", "task_001",
                "title", "Todo task",
                "status", "todo",
                "createdAt", "2026-03-18T12:00:00.000Z",
                "updatedAt", "2026-03-18T12:00:00.000Z"
            ),
            Map.of(
                "id", "task_002",
                "title", "Doing task",
                "status", "doing",
                "createdAt", "2026-03-18T12:10:00.000Z",
                "updatedAt", "2026-03-18T12:10:00.000Z"
            )
        )
    );
    testContext.packages.put("designSystem", new DesignSystemContract() {
      @Override
      public String feedback(String type, String message) {
        return message == null || message.isBlank() ? "" : "<div>" + message + "</div>";
      }

      @Override
      public String taskBoard(String columnsHtml) {
        return "<section>" + columnsHtml + "</section>";
      }

      @Override
      public String taskColumn(String title, int count, String itemsHtml) {
        return "<div>" + title + count + itemsHtml + "</div>";
      }

      @Override
      public String taskCard(String title, String description, String status, String actionsHtml) {
        return "<article>" + title + status + actionsHtml + "</article>";
      }

      @Override
      public String emptyColumnState(String message) {
        return "<div>" + message + "</div>";
      }
    });

    TaskListUiCase uiCase = new TaskListUiCase(testContext);
    String view = uiCase.view();
    if (view == null || view.isBlank()) {
      throw new IllegalStateException("test: view must return a visual unit");
    }

    TaskListDomainCase.TaskListOutput result = uiCase.load(new TaskListDomainCase.TaskListInput());
    if (result.tasks().size() != 2) {
      throw new IllegalStateException("test: ui service must return the mocked task list");
    }

    ViewState viewState = new ViewState();
    viewState.tasks = result.tasks();
    viewState.loading = false;
    ViewModel viewModel = uiCase._viewmodel(viewState);

    long todoCount = viewModel.columns.stream()
        .filter(column -> "todo".equals(column.status))
        .findFirst()
        .orElseThrow()
        .tasks
        .size();
    long doingCount = viewModel.columns.stream()
        .filter(column -> "doing".equals(column.status))
        .findFirst()
        .orElseThrow()
        .tasks
        .size();

    if (todoCount != 1 || doingCount != 1) {
      throw new IllegalStateException("test: ui viewmodel must group tasks by status");
    }
  }

  public TaskListDomainCase.TaskListOutput load(TaskListDomainCase.TaskListInput input) throws Exception {
    return _service(input);
  }

  protected ViewModel _viewmodel(ViewState currentState) {
    ViewModel model = new ViewModel();
    model.feedback = currentState.error;
    model.columns = List.of(
        column("todo", "To Do", currentState, currentState.tasks.stream()
            .filter(task -> "todo".equals(task.status()))
            .toList()),
        column("doing", "Doing", currentState, currentState.tasks.stream()
            .filter(task -> "doing".equals(task.status()))
            .toList()),
        column("done", "Done", currentState, currentState.tasks.stream()
            .filter(task -> "done".equals(task.status()))
            .toList())
    );
    return model;
  }

  protected TaskListDomainCase.TaskListOutput _service(
      TaskListDomainCase.TaskListInput input
  ) throws Exception {
    TaskListDomainCase domainCase = new TaskListDomainCase();
    domainCase.validate(input);
    return _repository(input);
  }

  protected TaskListDomainCase.TaskListOutput _repository(
      TaskListDomainCase.TaskListInput input
  ) throws Exception {
    Object response = resolveApiClient().request(Map.of(
        "method", "GET",
        "url", "/tasks"
    ));

    if (!(response instanceof Map<?, ?> mapResponse)) {
      throw new IllegalStateException("task_list.ui received an invalid list response");
    }

    @SuppressWarnings("unchecked")
    Map<String, Object> output = (Map<String, Object>) mapResponse;
    TaskListDomainCase.TaskListOutput result = TaskListDomainCase.TaskListOutput.fromMap(output);
    new TaskListDomainCase().validateOutput(result);
    return result;
  }

  private ColumnViewModel column(
      String status,
      String title,
      ViewState currentState,
      List<TaskListDomainCase.Task> tasks
  ) {
    ColumnViewModel column = new ColumnViewModel();
    column.status = status;
    column.title = title;
    column.tasks = tasks;
    column.emptyMessage = currentState.loading
        ? "Loading cards..."
        : "No cards in " + status + ".";
    return column;
  }

  private DesignSystemContract resolveDesignSystem() {
    Object value = ctx.packages.get("designSystem");
    if (!(value instanceof DesignSystemContract designSystem)) {
      throw new IllegalStateException("task_list.ui requires packages.designSystem");
    }
    return designSystem;
  }

  private RenderCardActions resolveRenderCardActions() {
    Object value = ctx.extra.get("renderCardActions");
    if (value == null) {
      return null;
    }

    if (!(value instanceof RenderCardActions renderCardActions)) {
      throw new IllegalStateException("task_list.ui requires a valid renderCardActions callback");
    }

    return renderCardActions;
  }

  private AppHttpClient resolveApiClient() {
    if (ctx.api == null) {
      throw new IllegalStateException("task_list.ui requires ctx.api");
    }

    return ctx.api;
  }

  private static final class NoopLogger implements AppLogger {
    @Override
    public void debug(String message, Map<String, Object> meta) {}

    @Override
    public void info(String message, Map<String, Object> meta) {}

    @Override
    public void warn(String message, Map<String, Object> meta) {}

    @Override
    public void error(String message, Map<String, Object> meta) {}
  }
}

File: examples/python/cases/tasks/task_list/task_list_ui_case.py
# ==========================================================================
# APP v1.1.4
# cases/tasks/task_list/task_list_ui_case.py
# --------------------------------------------------------------------------
# UI surface for the task_list Case.
#
# Responsibility:
# - present the task board as a three-column tkinter widget
# - group tasks into todo / doing / done columns via _viewmodel
# - fetch tasks from the backend API via _repository
# - pass the render_card_actions callback from ctx.extra to TaskBoard
#
# Canonical grammar:
#   view -> _viewmodel -> _service -> _repository
#
# Framework: tkinter
# Design system: packages/design_system consumed via ctx.packages
# ==========================================================================

from __future__ import annotations

from typing import Any, Callable, Literal

from core.ui_case import BaseUiCase, UiContext
from core.shared.app_structural_contracts import AppCaseError

# ==========================================================================
# Shared types
# --------------------------------------------------------------------------
# Cases are self-contained — types are redefined per file.
# ==========================================================================

TaskStatus = Literal["todo", "doing", "done"]
VALID_STATUSES = ("todo", "doing", "done")


class Task(dict):  # type: ignore[type-arg]
    """Runtime-typed task record (id, title, status, createdAt, updatedAt)."""


class TaskListInput(dict):  # type: ignore[type-arg]
    """Empty input — no filters accepted."""


class TaskListOutput(dict):  # type: ignore[type-arg]
    """Output containing a list of Task records."""


# ==========================================================================
# Column configuration
# --------------------------------------------------------------------------
# Ordered left-to-right: todo, doing, done.
# ==========================================================================

_COLUMN_CONFIG = [
    {
        "status_key": "todo",
        "title": "To Do",
        "empty_message": "No tasks in To Do",
    },
    {
        "status_key": "doing",
        "title": "Doing",
        "empty_message": "No tasks in Doing",
    },
    {
        "status_key": "done",
        "title": "Done",
        "empty_message": "No tasks in Done",
    },
]


# ==========================================================================
# TaskListUi
# ==========================================================================

class TaskListUi(BaseUiCase):
    """UI surface for the task_list Case."""

    # ==================================================================
    # Required: view
    # ==================================================================

    def view(self) -> Any:
        """
        Public entrypoint — returns the TaskBoard tkinter widget.

        Fetches tasks via _service(), groups them via _viewmodel(), then
        renders a TaskBoard with the render_card_actions callback from
        ctx.extra (if provided by the host).
        """
        packages = self.ctx.get("packages", {})
        design_system: dict[str, Any] = packages.get("design_system") or {}
        task_board_cls = design_system.get("TaskBoard")

        if task_board_cls is None:
            raise AppCaseError(
                "INTERNAL",
                "task_list.ui requires packages['design_system']['TaskBoard']",
            )

        tasks: list[dict[str, Any]] = self._service()
        columns = self._viewmodel(tasks)

        render_card_actions: Callable[..., Any] | None = (
            self.ctx.get("extra", {}).get("render_card_actions")
        )

        # TaskBoard requires a tkinter parent; acquire it from ctx.renderer
        # when available, otherwise fall back to a bare tk.Frame stub for
        # environments (tests) that do not provide a renderer.
        renderer = self.ctx.get("renderer")

        board = task_board_cls(
            master=renderer,
            columns=columns,
            render_card_actions=render_card_actions,
        )
        return board

    # ==================================================================
    # Internal slots
    # ==================================================================

    def _viewmodel(self, tasks: list[dict[str, Any]]) -> list[dict[str, Any]]:  # type: ignore[override]
        """
        Transform a flat task list into three column descriptors.

        Groups tasks by status and attaches column titles and empty messages.
        The task order within each column preserves the input order (which
        is expected to be createdAt descending as returned by the API).
        """
        grouped: dict[str, list[dict[str, Any]]] = {
            "todo": [],
            "doing": [],
            "done": [],
        }
        for task in tasks:
            status = task.get("status", "todo")
            if status in grouped:
                grouped[status].append(task)

        columns: list[dict[str, Any]] = []
        for cfg in _COLUMN_CONFIG:
            key = cfg["status_key"]
            columns.append(
                {
                    "status_key": key,
                    "title": cfg["title"],
                    "empty_message": cfg["empty_message"],
                    "tasks": grouped.get(key, []),
                }
            )

        return columns

    def _service(self) -> list[dict[str, Any]]:  # type: ignore[override]
        """
        Fetch tasks from the backend via _repository().

        Returns the list of tasks (may be empty).
        If the fetch fails, returns an empty list and logs the error.
        """
        try:
            result: TaskListOutput = self._repository()
            return result.get("tasks", [])  # type: ignore[return-value]
        except Exception as exc:
            logger = self.ctx.get("logger")
            if logger is not None:
                logger.error(
                    "task_list_ui: failed to fetch tasks",
                    {"error": str(exc)},
                )
            return []

    def _repository(self) -> TaskListOutput:  # type: ignore[override]
        """
        Fetch all tasks from the backend API.

        Calls GET /tasks via ctx.api.request() and returns the unwrapped
        TaskListOutput. The API adapter is expected to unwrap the {success,
        data} envelope and return data directly (or raise on failure).
        """
        api = self.ctx["api"]
        data: Any = api.request({"method": "GET", "path": "/tasks"})

        # The adapter may return the bare tasks list or the full output envelope.
        if isinstance(data, dict) and "tasks" in data:
            return data  # type: ignore[return-value]

        # Gracefully handle adapters that return the tasks list directly.
        if isinstance(data, list):
            return {"tasks": data}  # type: ignore[return-value]

        return {"tasks": []}  # type: ignore[return-value]

    # ==================================================================
    # Test
    # ==================================================================

    def test(self) -> None:
        """
        Minimal structural test for the UI surface.

        Verifies:
        - the class can be instantiated with a minimal UiContext
        - view(), _viewmodel(), _service(), _repository() exist as methods
        - _viewmodel correctly groups tasks into 3 columns
        - _viewmodel assigns correct empty messages when columns are empty
        """
        # Verify all required methods exist.
        assert callable(getattr(self, "view", None)), "view() must exist"
        assert callable(getattr(self, "_viewmodel", None)), "_viewmodel() must exist"
        assert callable(getattr(self, "_service", None)), "_service() must exist"
        assert callable(getattr(self, "_repository", None)), "_repository() must exist"

        # ----------------------------------------------------------------
        # _viewmodel: empty input produces 3 columns with empty tasks
        # ----------------------------------------------------------------
        columns = self._viewmodel([])
        assert len(columns) == 3, "_viewmodel must return exactly 3 columns"

        status_keys = [col["status_key"] for col in columns]
        assert status_keys == ["todo", "doing", "done"], (
            "Columns must be ordered: todo, doing, done"
        )

        for col in columns:
            assert col["tasks"] == [], (
                f"Column '{col['status_key']}' must have empty tasks for empty input"
            )
            assert col["empty_message"], (
                f"Column '{col['status_key']}' must have a non-empty empty_message"
            )

        # ----------------------------------------------------------------
        # _viewmodel: tasks are grouped into the correct columns
        # ----------------------------------------------------------------
        sample_tasks: list[dict[str, Any]] = [
            {
                "id": "t1",
                "title": "Todo task",
                "status": "todo",
                "createdAt": "2024-01-03T00:00:00Z",
                "updatedAt": "2024-01-03T00:00:00Z",
            },
            {
                "id": "t2",
                "title": "Doing task",
                "status": "doing",
                "createdAt": "2024-01-02T00:00:00Z",
                "updatedAt": "2024-01-02T00:00:00Z",
            },
            {
                "id": "t3",
                "title": "Done task",
                "status": "done",
                "createdAt": "2024-01-01T00:00:00Z",
                "updatedAt": "2024-01-01T00:00:00Z",
            },
        ]

        grouped_columns = self._viewmodel(sample_tasks)
        todo_col = next(c for c in grouped_columns if c["status_key"] == "todo")
        doing_col = next(c for c in grouped_columns if c["status_key"] == "doing")
        done_col = next(c for c in grouped_columns if c["status_key"] == "done")

        assert len(todo_col["tasks"]) == 1, "todo column must have 1 task"
        assert todo_col["tasks"][0]["id"] == "t1", "todo column must contain t1"

        assert len(doing_col["tasks"]) == 1, "doing column must have 1 task"
        assert doing_col["tasks"][0]["id"] == "t2", "doing column must contain t2"

        assert len(done_col["tasks"]) == 1, "done column must have 1 task"
        assert done_col["tasks"][0]["id"] == "t3", "done column must contain t3"

        # ----------------------------------------------------------------
        # _viewmodel: empty messages
        # ----------------------------------------------------------------
        assert "To Do" in todo_col["empty_message"], (
            "todo empty_message must reference 'To Do'"
        )
        assert "Doing" in doing_col["empty_message"], (
            "doing empty_message must reference 'Doing'"
        )
        assert "Done" in done_col["empty_message"], (
            "done empty_message must reference 'Done'"
        )
